commons/src/main/java/org/infinispan/commons/api/functional/FunctionalMap.java

```diff
@@ -0,0 +1,634 @@
+       * {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as
+       * {@link javax.cache.Cache#removeAll(Set)}.
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>It makes sense to expose global operation like this
+       *    instead of forcing users to iterate over the keys to lookup and call
+       *    get individually since Infinispan can do things more efficiently.
+       *    </li>
+       *    <li>Since this is a write-only operation, no entry attributes can be
+       *    queried, hence the only reasonable thing can be returned is Void.
+       *    </li>
+       *    <li>It still makes sense to return a CloseableIterator<Void> instead of
+       *    CompletableFuture<Void> in case the user wants to do something as each
+       *    operation completes, e.g. keep track of progress.
+       *    </li>
+       * </ul>
+       *
+       * @param keys the keys associated with each of the {@link WriteEntryView}
+       *             passed in the function callbacks
+       * @param f operation that the {@link WriteEntryView} associated with
+       *          one of the keys passed in
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalMany(Set<? extends K> keys, Consumer<WriteEntryView<V>> f);
+
+      /**
+       * Evaluate a write-only {@link Consumer} operation with the
+       * {@link WriteEntryView} of the value associated with the key, for all
+       * existing keys in functional map, and returns a
+       * {@link CloseableIterator} that allows navigating completion
+       * of the operation executions.
+       *
+       * This method can be used to implement operations such as
+       * {@link javax.cache.Cache#removeAll()}.
+       *
+       * DESIGN RATIONALE:
+       * <ul>
+       *    <li>In the original design, this method would have taken no
+       *    parameters and returned a CloseableIterator<WriteEntryView<V>>,
+       *    but doing so would have created encapsulation problems, since any
+       *    changes to the entry view would need to be committed, and that
+       *    would happen outside of the interceptor stack scope. To avoid this
+       *    problem, simply pass in a consumer with the desired operation and
+       *    then, the interceptor stack can control how the consumer gets
+       *    executed and make sure it's committed.</li>
+       * </ul>
+       *
+       * @param f operation that the {@link WriteEntryView} associated with
+       *          one of the keys passed in
+       * @return a {@link CloseableIterator} that can be navigated to follow progress
+       *         as operations get invoked
+       */
+      CloseableIterator<Void> evalAll(Consumer<WriteEntryView<V>> f);
```

Comment:
@galderz I think required annotations are just as much a part of the API as regular interfaces and methods are :)

BTW, the value matcher is specifically because we cannot provide guarantees around the result of conditional commands when key ownership changes. Sometimes the result of the operation will be wrong, but at least we can ensure that the entry is updated to the same value on all the owners (unless the originator crashes). I have no idea how that would work for generic lambdas.